Add AWS Fargate profile support

ferruzzi opened the pull request:
Adds hooks, operators, tests, docs, and example DAGs to support AWS Fargate profiles.  

- Includes changes to the existing EKS modules to incorporate the new hooks and operators.
- Bumps the minimum `moto` version so we consume the new Fargate support there.
- Completes the work started in  https://github.com/apache/airflow/pull/16571 and https://github.com/apache/airflow/pull/17951 for adding better EKS support.

- Passes local pre-commit tests and linters.

related: https://github.com/apache/airflow/pull/16571
related: https://github.com/apache/airflow/pull/17951

Commit:
adding hook and operators for eks fargate profile apis

ferruzzi committed:
Fargate Docs

ferruzzi committed:
Catchup commit

Bring Vishal's work up to our current understanding and patterns.

ferruzzi committed:
Fargate Hook Tests

ferruzzi committed:
Fargate Operator Tests

ferruzzi committed:
Add Fargate support to "Create All" logic

ferruzzi committed:
Add Fargate support to Delete Cluster Operator's `force` logic

ferruzzi committed:
Refactor EKS Hook and Operator

ferruzzi committed:
Doc formatting fixes

Co-authored-by:  o-nikolas <onikolas@amazon.com>

mik-laj commented on airflow/providers/amazon/aws/sensors/eks.py:
```diff
@@ -77,6 +82,65 @@ def poke(self, context):
+            clusterName=self.cluster_name, fargateProfileName=self.fargate_profile_name
+        )
+        self.log.info("Fargate profile state: %s", fargate_profile_state)
+        return fargate_profile_state == self.target_state
```
What if the operator reaches the terminal state before reaching the expected state, eg CREATE_FAILED? In that case, we should raise a exception and stop waiting because the state will never change.

ferruzzi replied:
That's fair.   What are you thinking?   A `List` of terminal states and a quick `if current state is terminal and not the target then throw an exception`?

mik-laj replied:
We have to check whether we have terminal the end states. If so, raise an AirflowException to stop the sensor.

````
TERMINAL_STATES = {CREATE_FAILED, DELETE_FAILED, NONEXISTENT}
````
````
current_terminal_states = TERMINAL_STATES - {self.target_state}
if fargate_profile_state in current_terminal_states:
    raise AirflowException("Terminal state reached. Currentt state: {fargate_profile_state}, Excepted state: {self.target_state}")
return fargate_profile_state == self.target_state
````

ferruzzi replied:
Addressed in https://github.com/apache/airflow/pull/18645/commits/575dffa2b252a004efa4676e98dbb6add2f686a7

josh-fell commented on airflow/providers/amazon/aws/example_dags/example_eks_with_fargate_in_one_step.py:
```diff
@@ -0,0 +1,100 @@
+    start_date=days_ago(2),
+    max_active_runs=1,
+    tags=['example'],
+) as dag:
```
We should have new example DAGs use a static `start_date` value.  There is an ongoing cleanup effort across all example DAGs to transition away from using `days_ago(n)`.

ferruzzi replied:
Sure.  What should the static start date be set to?

josh-fell replied:
It doesn't matter really; we've been using `datetime(2021, 1, 1)`.

ferruzzi replied:
addressed in https://github.com/apache/airflow/pull/18645/commits/a8a1a41a7a4e345b03d1b485ec8b0aaedd3dd729

josh-fell commented on airflow/providers/amazon/aws/example_dags/example_eks_with_fargate_in_one_step.py:
```diff
@@ -0,0 +1,100 @@
+    # Create an Amazon EKS cluster control plane and an AWS Fargate compute platform in one step.
+    create_cluster_and_fargate_profile = EKSCreateClusterOperator(
+        task_id='create_eks_cluster_and_fargate_profile',
+        cluster_name=CLUSTER_NAME,
```
Since `cluster_name` is used in every operator with the same value in this DAG, would be a great use case to add the arg to `default_args`.

You may need to include a comment that this arg is set in `default_args` of the DAG so it's known while reading the operator doc.

ferruzzi replied:
Addressed in https://github.com/apache/airflow/pull/18645/commits/a8a1a41a7a4e345b03d1b485ec8b0aaedd3dd729

josh-fell commented on airflow/providers/amazon/aws/example_dags/example_eks_with_fargate_profile.py:
```diff
@@ -0,0 +1,129 @@
+}
+
+
+with DAG(
```
Same `start_date` and `cluster_name` comments are also applicable here.

ferruzzi replied:
addressed in https://github.com/apache/airflow/pull/18645/commits/a8a1a41a7a4e345b03d1b485ec8b0aaedd3dd729

ferruzzi committed:
Sample DAG updates

Per PR suggestions:
- Updated to static start dates
- Moved `cluster_name` into `default_args`

mik-laj commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -23,25 +23,41 @@
+
+     - If argument 'compute' is provided with a value of 'nodegroup', will also
+         attempt to create an Amazon EKS Managed Nodegroup for the cluster.
+         See EKSCreateNodegroupOperator documentation for requirements.
```
Can you use  `:class:` role here?

ferruzzi replied:
Should be addressed in https://github.com/apache/airflow/pull/18645/commits/b577f159075d1c6b75d90c54303041b78025f122

mik-laj commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -23,25 +23,41 @@
+
+    -  If argument 'compute' is provided with a value of 'fargate', will also attempt to create an AWS
+         Fargate profile for the cluster.
+         See EKSCreateFargateProfileOperator documentation for requirements.
```
Can you use  `:class:` role here?

ferruzzi replied:
Should be addressed in https://github.com/apache/airflow/pull/18645/commits/b577f159075d1c6b75d90c54303041b78025f122

mik-laj commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -154,11 +204,18 @@ def execute(self, context):
+                fargateProfileName=self.fargate_profile_name,
+                podExecutionRoleArn=self.fargate_pod_execution_role_arn,
+                selectors=self.fargate_selectors,
+            )
```
Should we raise an exception on an invalid type?

ferruzzi replied:
Covered by https://github.com/apache/airflow/pull/18645#discussion_r719868145

mik-laj commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -93,21 +122,42 @@ def __init__(
+                            compute=FARGATE_FULL_NAME, requirement='fargate_pod_execution_role_arn'
+                        )
+                    )
+            except KeyError:
```
When this situation can happen?

ferruzzi replied:
This needs to get changed to an `else`.  In a previous iteration there was an assignment in there that could fail and I missed this on the refactor.   

TODO:  Drop the `try` and make the `except` a `check if it is a valid compute value`.

ferruzzi replied:
Should be addressed in https://github.com/apache/airflow/pull/18645/commits/b577f159075d1c6b75d90c54303041b78025f122

mik-laj commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -227,6 +284,78 @@ def execute(self, context):
+        self.pod_execution_role_arn = pod_execution_role_arn
+        self.selectors = selectors
+        self.fargate_profile_name = (
+            fargate_profile_name or f'{cluster_name}{datetime.now().strftime("%Y%m%d_%H%M%S")}'
```
Operator attributes should be static. Otherwise, the DAG will be serialized multiple times to the database.

ferruzzi replied:
Is the solution to use  `self.fargate_profile_name = fargate_profile_name` here in the `init`, and  then below in the `execute()` method add the `or foo` portion to the hook call?

```
    eks_hook.create_fargate_profile(
        clusterName=self.cluster_name,
         fargateProfileName=self.fargate_profile_name or f'{self.cluster_name}{datetime.now().strftime("%Y%m%d_%H%M%S")}',
        podExecutionRoleArn=self.pod_execution_role_arn,
        selectors=self.selectors,
    )
 ```

ferruzzi replied:
I believe this is addressed by  https://github.com/apache/airflow/pull/18645/commits/b577f159075d1c6b75d90c54303041b78025f122

mik-laj commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -280,41 +409,83 @@ def execute(self, context):
+        nodegroups = eks_hook.list_nodegroups(clusterName=self.cluster_name)
+        nodegroup_count = len(nodegroups)
+        if nodegroup_count:
+            self.log.info(CAN_NOT_DELETE_MSG.format(compute=NODEGROUP_FULL_NAME, count=nodegroup_count))
```
```suggestion
        if nodegroups:
            self.log.info(CAN_NOT_DELETE_MSG.format(compute=NODEGROUP_FULL_NAME, count=len(nodegroups)))
```
More pythonic approach.

ferruzzi replied:
Should be addressed in https://github.com/apache/airflow/pull/18645/commits/b577f159075d1c6b75d90c54303041b78025f122

mik-laj commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -280,41 +409,83 @@ def execute(self, context):
+        fargate_profiles = eks_hook.list_fargate_profiles(clusterName=self.cluster_name)
+        fargate_profile_count = len(fargate_profiles)
+        if fargate_profile_count:
+            self.log.info(CAN_NOT_DELETE_MSG.format(compute=FARGATE_FULL_NAME, count=fargate_profile_count))
```
```suggestion
        if fargate_profiles:
            self.log.info(CAN_NOT_DELETE_MSG.format(compute=FARGATE_FULL_NAME, count=len(fargate_profiles)))
```

ferruzzi replied:
Should be addressed in https://github.com/apache/airflow/pull/18645/commits/b577f159075d1c6b75d90c54303041b78025f122

mik-laj commented on setup.py:
```diff
@@ -513,7 +513,7 @@ def write_version(filename: str = os.path.join(*[my_dir, "airflow", "git_version
     'jsondiff',
     'mongomock',
-    'moto~=2.2, >=2.2.7',
+    'moto~=2.2, >=2.2.4.dev12',
```
moto 2..2.7 is required by kinesis. See: https://github.com/apache/airflow/pull/18337

ferruzzi replied:
gah.   I messed up a merge there, I'll correct it back to `2.2.7`

ferruzzi replied:
Should be addressed in https://github.com/apache/airflow/pull/18645/commits/b577f159075d1c6b75d90c54303041b78025f122

josh-fell commented on airflow/providers/amazon/aws/example_dags/example_eks_with_nodegroup_in_one_step.py:
```diff
@@ -36,11 +36,15 @@
 
 with DAG(
     dag_id='example_eks_using_defaults_dag',
+    default_args=default_args,
```
Can you add `default_args={'cluster_name': CLUSTER_NAME},` directly here as well?

ferruzzi replied:
Should be addressed in https://github.com/apache/airflow/pull/18645/commits/b577f159075d1c6b75d90c54303041b78025f122

josh-fell commented on airflow/providers/amazon/aws/example_dags/example_eks_with_fargate_profile.py:
```diff
@@ -0,0 +1,124 @@
+
+with DAG(
+    dag_id='example_eks_with_fargate_profile_dag',
+    default_args=default_args,
```
It's more readable to have the `cluster_name` arg directly set in the `DAG()` instantiation.  Another instance of something else being cleaned up from old DAG examples.  Sorry for the nit here.

```
with DAG(
    dag_id='example_eks_with_fargate_profile_dag',
    default_args={'cluster_name': CLUSTER_NAME},
  ....
)
```

ferruzzi replied:
Does this work with templated values as well?  ie:  

```
with DAG(
    dag_id='example_eks_with_fargate_profile_dag',
    default_args={'cluster_name': "{{ dag_run.conf['cluster_role_arn'] }}"},
  ....
)
```

josh-fell replied:
Sure does!

ferruzzi replied:
Should be addressed in https://github.com/apache/airflow/pull/18645/commits/b577f159075d1c6b75d90c54303041b78025f122

josh-fell commented on airflow/providers/amazon/aws/example_dags/example_eks_with_nodegroups.py:
```diff
@@ -40,11 +39,15 @@
 
 with DAG(
     dag_id='example_eks_with_nodegroups_dag',
+    default_args=default_args,
```
Same comment here too. Thanks for adding these in! 🚀

ferruzzi replied:
Should be addressed in https://github.com/apache/airflow/pull/18645/commits/b577f159075d1c6b75d90c54303041b78025f122

ferruzzi committed:
Multiple small PR fixes

ferruzzi committed:
Add Terminal State checks and unit tests to Sensors

ferruzzi committed:
Update Example DAGS to meet deprecation notices

potiuk reviewed (approved):
LGTM - but I would love @mik-laj to confirm that all is addressed. I'd love to include that in the adhoc amazon provider.
